deduce.go

```diff
@@ -690,6 +695,253 @@ func (sm *SourceMgr) deduceFromPath(path string) (deductionFuture, error) {
 	}, nil
 }
 
+type deductionCoordinator struct {
+	ctx      context.Context
```

Comment:
> we could thread a context through that as well
Yep! That's exactly what I'd expect to see naively. 

> But while that's a fine guarantee for an implementation to provide, within gps, we have no idea if the caller has wired the long-lived context together with the ephemeral call contexts. To ensure correctness, we still have to internally combine cancelation.

I don't mean to be dense, but I'm still not totally getting this. If the calls are coming from the solver, not `dep` itself, then we seem to be sure of the provenance of the contexts, right? We know they came as children made by the solver off of the original one from `dep` (or some other implementation using the solver).

Are you concerned that there are multiple entrypoints to this code, and you want to be sure the caller is using the same context for them? If so, I think the right approach is to shrink the entrypoints. Do you have a concrete case here? That might help me understand.

> Yes, and this is mostly the counterargument that Peter made on reddit. (I also did up what you're describing in some pseudocode).

Hm, this pseudocode isn't what I was thinking of. There's no need for quitchan there - you'd just call `cancel()` at the end of the function directly. Each spawned goroutine is checking whether the context is closed before (or while) it does significant work like network calls. If you want to only exit once all spawned goroutines have exited, you keep track of that with `sync.WaitGroup`.

I agree that both approaches eventually work, of course. No question there! This is more about how easy those approaches are to keep up and maintain. I just think it's simpler to think about the one flow of contexts than this idea of merging two.